pkg/apis/apis/v1alpha1/types_apiresourceschema.go

```diff
@@ -163,3 +163,95 @@ func (v *APIResourceVersion) SetSchema(schema *apiextensionsv1.JSONSchemaProps)
 	v.Schema.Raw = raw
 	return nil
 }
+
+// +crd
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Cluster,categories=kcp
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+
+// APIConversion contains rules to convert between different API versions in an APIResourceSchema. The name must match
+// the name of the APIResourceSchema for the conversions to take effect.
+type APIConversion struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec holds the desired state.
+	Spec APIConversionSpec `json:"spec"`
+}
+
+// APIConversionSpec contains rules to convert between different API versions in an APIResourceSchema.
+type APIConversionSpec struct {
+	// +required
+	// +listType=map
+	// +listMapKey=from
+	// +listMapKey=to
+
+	// conversions specify rules to convert between different API versions in an APIResourceSchema.
+	Conversions []APIVersionConversion `json:"conversions"`
+}
+
+// APIVersionConversion contains rules to convert between two specific API versions in an
+// APIResourceSchema. Additionally, to avoid data loss when round-tripping from a version that
+// contains a new field to one that doesn't and back again, you can specify a list of fields to
+// preserve (these are stored in annotations).
+type APIVersionConversion struct {
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=^v[1-9][0-9]*([a-z]+[1-9][0-9]*)?$
+
+	// from is the source version.
+	From string `json:"from"`
+
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=^v[1-9][0-9]*([a-z]+[1-9][0-9]*)?$
+
+	// to is the target version.
+	To string `json:"to"`
+
+	// +required
+	// +listType=map
+	// +listMapKey=destination
+
+	// rules contains field-specific conversion expressions.
+	Rules []APIConversionRule `json:"rules"`
+
+	// +optional
+
+	// preserve contains a list of JSONPath expressions to fields to preserve in the originating version
+	// of the object, relative to its root, such as '.spec.name.first'.
+	Preserve []string `json:"preserve,omitempty"`
+}
+
+// APIConversionRule specifies how to convert a single field.
+type APIConversionRule struct {
+	// +required
+
+	// field is a JSONPath expression to the field in the originating version of the object, relative to its root, such
+	// as '.spec.name.first'.
+	Field string `json:"field"`
+
+	// +required
+
+	// destination is a JSONPath expression to the field in the target version of the object, relative to
+	// its root, such as '.spec.name.first'.
+	Destination string `json:"destination"`
+
+	// +optional
+
+	// transformation is an optional CEL expression used to execute user-specified rules to transform the
+	// originating field -- identified by 'self' -- to the destination field.
+	Transformation string `json:"transformation,omitempty"`
+}
```

Comment:
am still thinking we are missing something here:

1. assume destination `a.b[*].c` and field `a.b[*]`. Then we can access e.g. a.b[*].old_c` to derive `a.b[42].c`. That's fine. So I guess field must be a prefix of destination.
2. assume destination `a.b[*].c` and field `a.b`. Then we can access e.g. `a.b[42].old_c` to derive `a.b[42].c`, i.e. copy a static value into each item in the list.
3. assume destination `a.b[*].c` and field `a.b`. Then we can access e.g. a.b.old_c` to derive `a.b[42].c`, but we cannot reference `a.b[42].d` at the same time. **We could map `root, a, a.b, a.b[42]` into the CEL context to give access to every level. Any better idea**?
4. we would also like to access the index, or in case of maps the key. We might also want to find a way to put that in the CEL context.

Reply:
This PR doesn't support indexing currently

Reply:
As discussed on Meet, the API in the PR is ok to merge, and we can do follow-ups to evolve to support more advanced use cases.